app/src/main/java/org/simple/clinic/util/Distance.kt

```diff
@@ -0,0 +1,15 @@
+package org.simple.clinic.util
+
+sealed class Distance {
+  abstract fun meters(): Double
+  abstract fun kilometers(): Double
```

Comment:
`meters()` isn't needed yet, but I wanted to show an example of what we can have in the future apart from just `kilometers()`.

Reply:
I actually think a better way to do it is how `java.util.time.Duration` is designed:

```kotlin
data class Distance(val distance: Long, val unit: DistanceUnit) {
  private val millimeters = distance * unit.millimeterConversionFactor

  companion object {
    fun ofMillimeters(millimeters: Long) = Distance(millimeters, MILLIMETERS)
    fun ofMeters(meters: Long) = Distance(millimeters, METERS)
  }

  enum class DistanceUnit(val millimeterConversionFactor: Long) {
    MILLIMETERS(1),
    METERS(1_000),
    KILOMETERS(1_000_000)
  }
}
```

This will enable us to do things like adding two `Distance` instances, converting from one to another easily, etc.

Reply:
That was exactly my intention, but I wanted to use sealed classes instead of enums. The units can still be converted from one to another this way. I'll try improving this.

Reply:
Why though? Sealed classes are not a good fit for this use case. They are good when you want to have a single way to define a limited set of types that are related, but this doesn't seem like one of those cases.

Reply:
I like using sealed classes even when the properties of each type is the same. If enums are used in this case, we'll have to use two different values for representing them:

```kotlin
value = 15
unit = DistanceUnit.KILOMETER
```

With classes,

```kotlin
distance = Kilometer(15.0)
```

With enums, when returning them from a function, we'll have to combine them in some manner:

```kotlin
fun haversine(): Pair<Double, DistanceUnit> {
  return 15.0 to DistanceUnit.KILOMETER
}
```

With classes,

```kotlin
fun haversine(): Distance {
  return Kilometer(15.0)
}
```

For converting them from one unit to another, you'll have to do:

```kotlin
val (value, unit) = haversine(...)
val meters = unit.toMeters(value)
```

With classes, 

```kotlin
val distance = haversine(...)
val meters = distance.meters()
```

I don't think `TimeUnit` is a good reference for design. Using them introduces a lot of verbosity. `Duration` from Java 8 date-time is neater.

Reply:
Wait, I looked at the source of `Duration`. Your example is not how it's designed. I actually thought you were referencing `TimeUnit` which maintains the unit and the value separately.

Reply:
Is there any specific reason why you prefer referencing distance units through `Distance.of...()`? I like the idea of being able to create objects of each distance unit separately

```kotlin
fun haversine(): Distance {
  return Distance.ofKilometers(15.0)
}
```

vs,

```kotlin
fun haversine(): Distance {
  return Kilometers(15.0)
}
```

I understand your reason for easier conversion. I'll add a `meters` field to the base class so each unit subclass doesn't have to deal with conversion.

Reply:
> Wait, I looked at the source of `Duration`. Your example is not how it's designed. I actually thought you were referencing `TimeUnit` which maintains the unit and the value separately.

Oh yeah, I just meant the entire api style, like using `Distance.ofMeters()`, and not the actual implementation itself.

Reply:
Another reason for having a single class that uses the same values to represent it is we can make this `Distance` class `Comparable`. Doing this with sealed classes will be cumbersome. Representing this in this way also makes (de)serializing `Distance` instances very easy. Otherwise, you'll need to have a polymorphic (de)serializer of some sort that will have a growing like of pointless `when` statements.

Reply:
> Another reason for having a single class that uses the same values to represent it is we can make this `Distance` class `Comparable`.

That should still work with subclasses because the raw value will stay inside `Distance`. In fact, there's already a `compareTo()` function. We just need to mark the class as `Comparable`.

> Representing this in this way also makes (de)serializing `Distance` instances very easy. Otherwise, you'll need to have a polymorphic (de)serializer of some sort that will have a growing like of pointless `when` statements.

True, but let's worry about this when we reach there? I know that Moshi is already working on adding support for sealed classes so we won't have to deal with this very soon. https://github.com/square/moshi/pull/741

Reply:
For the record: I'll change `Distance` to use factory functions for creating different units instead of using subclasses. 

sad_parrot.gif